lib/mongo/collection.rb

```diff
@@ -75,6 +75,55 @@ def initialize(database, name, options = {})
       @options = options.freeze
     end
 
+    # Get the read preference on this collection.
+    #
+    # @example Get the read preference.
+    #   collection.read_preference
+    #
+    # @return [ Mongo::ServerSelector ] The read preference.
+    #
+    # @since 2.0.0
+    def read_preference
+      @read_preference ||= options[:read] ? ServerSelector.get((options[:read]).merge(options)) :
+        database.read_preference
+    end
+
+    # Get the write concern on this collection.
+    #
+    # @example Get the write concern.
+    #   collection.write_concern
+    #
+    # @return [ Mongo::WriteConcern ] The write concern.
+    #
+    # @since 2.0.0
+    def write_concern
+      @write_concern ||= options[:write] ? WriteConcern.get(options[:write]) :
+        database.write_concern
+    end
+
+    # Provides a new collection with either a new read preference or new write concern
+    # merged over the existing read preference / write concern.
+    #
+    # @example Get a collection with changed read preference.
+    #   collection.with(:read => { :mode => :primary_preferred })
+    #
+    # @example Get a collection with changed write concern.
+    #   collection.with(:write => { w:  3 })
+
+    # @param [ Hash ] new_options The new options to use.
+    #
+    # @return [ Mongo::Collection ] A new collection instance.
+    #
+    # @since 2.1.0
+    def with(new_options)
```

Comment:
You could still merge the client options in the single line as well. I think if we provide a `#with` method just like `Mongo::Client` it needs to accept any options, not just read and write options.